File: Cpp/_225.cpp
// 225: Implement Stack with 2 Queues
// Need only 1 queue. 2 queues use same time complexity

// For push: push x, then repush those that were
// pushed before x. O(n) push and O(1) peek  

class MyStack {
    queue<int> queue;
public:
    MyStack() {}
    
    void push(int x) {
        queue.push(x);
        for (int i = 1, n = queue.size(); i < n; i++) {
            queue.push(pop()); // Move front e to back
        }
    }
    
    int pop() {
        int x = queue.front();
        queue.pop();
        return x;
    }
    
    int top() {
        return queue.front();
    }
    
    bool empty() {
        return queue.empty();
    }
};

File: Python3/_225.py
# 225: Implement Stack with 2 Queues
# Need only 1 queue. 2 queues use same time complexity

# For push: push x, then repush those that were
# pushed before x. O(n) push and O(1) peek  
class MyStack:
    def __init__(self):
        # As queue is 1-ended, choose only 1 pair:
        # (append + popleft) or (appendleft + pop) 
        self.queue = deque()
        
    def push(self, x: int) -> None:
        self.queue.append(x)
        for i in range(1, len(self.queue)):
            self.queue.append(self.queue.popleft())

    def pop(self) -> int:
        return self.queue.popright()

    def top(self) -> int:
        return self.queue[0]

    def empty(self) -> bool:
        return not bool(self.queue)
        